### PR TITLE
add waitlist notifications

### PR DESCRIPTION
**Pull request by upacyxou**
- add waitlist notifications

### CODE REVIEW

File: src/controllers/token.ts [hunk 1]
```diff
@@ -31,6 +31,36 @@ export default class TokenController {
...
+    @Ctx() ctx: Context
+  ) {
+    const { token } = body
+    const existingToken = await TokenModel.findOne({ token })
+    if (existingToken) {
+      existingToken.waitlist = true
+      await existingToken.save()
+      return { success: true }
+    }
+    if (!token) return ctx.throw(badRequest('No token'))
```
Comment (smyzin) [inline src/controllers/token.ts:46]: It's better to move this line up, to the L.40

File: src/controllers/token.ts [hunk 2]
```diff
@@ -31,6 +31,36 @@ export default class TokenController {
...
+    const newToken = new TokenModel({ token, waitlist: true })
+    await newToken.save()
+    return { success: true }
+  }
+
+  @Put('/waitlist/:token')
+  async markAsReceived(
+    @Params('token') notificationToken: string,
```
Comment (smyzin) [inline src/controllers/token.ts:54]: I get the idea, but seems like it's better to name it `token`, instead of `notificationToken` 🤔

**Review by smyzin — APPROVED** [review]

File: src/helpers/sendFirebaseNotification.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ interface FirebaseNotificationProps {
   body?: string
   postId?: number
   feedId?: number
-  type?: 'entanglement' | 'newpost'
+  type?: 'entanglement' | 'newpost' | 'waitlist'
```
Comment (smyzin) [inline src/helpers/sendFirebaseNotification.ts:15]: I assume it's better to create enum for that, imo.

File: src/controllers/token.ts [hunk 4]
```diff
@@ -31,6 +31,33 @@ export default class TokenController {
     }
   }
 
+  @Post('/waitlist')
+  async addWaitlistToken(
+    @Body({ required: true }) body: TokenWithSettings,
+    @Ctx() ctx: Context
+  ) {
+    const { token } = body
+    if (!token) return ctx.throw(badRequest('No token'))
```
Comment (T-Damer) [inline src/controllers/token.ts:40]: You'll always have token, validator won't accept body without it, we don't need this check

File: src/helpers/sendFirebaseNotification.ts (hunk 3 above)
Comment (T-Damer) [reply src/helpers/sendFirebaseNotification.ts:15]: enum is pretty the same in case we use those, no need for it

**Review by T-Damer — APPROVED** [review]

**Pull request merged**